### PR TITLE
Support BBC News (bbc.com/news)

### PR DESCRIPTION
**Pull request by atomicdryad**
bbc.com/news uses the same mediaselector format, but a different mediaselector url, and inlined json + xml playlist.sxml methods for obtaining video ids.

_download_media_selector is mostly redundant except for the hardcoded url; perhaps it should be made a third argument?

### CODE REVIEW

File: docs/supportedsites.md [hunk 1]
```diff
@@ -50,6 +50,7 @@
  - **Bandcamp**
  - **Bandcamp:album**
  - **bbc.co.uk**: BBC iPlayer
+ - **bbc.com**: BBC news videos
```
Comment (yan12125) [inline docs/supportedsites.md]: This file is updated automatically in releases. Please do not modify it in pull requests.

File: youtube_dl/extractor/bbcnews.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+            'title': 'Aerial footage showed the site of the crash in the Alps - courtesy BFM TV',
+            'description': 'Germanwings plane crash site in aerial video - Aerial footage showed the site of the crash in the Alps - courtesy BFM TV',
+            'duration': 47,
+        },
+        'params': {
+            'skip_download': True,
+        }
+    }]
+
+    def _duration_str2int(self, str):
```
Comment (yan12125) [inline youtube_dl/extractor/bbcnews.py]: I haven't tested it thoroughly, but maybe [parse_duration](https://github.com/rg3/youtube-dl/blob/master/youtube_dl/utils.py#L1299) have the same functionality.

File: docs/supportedsites.md (hunk 1 above)
Comment (atomicdryad) [inline docs/supportedsites.md]: Reverted

File: youtube_dl/extractor/bbcnews.py (hunk 2 above)
Comment (atomicdryad) [inline youtube_dl/extractor/bbcnews.py]: Yeah it does, removed

File: youtube_dl/extractor/bbcnews.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+    def _real_extract(self, url):
+        list_id = self._match_id(url)
+        webpage = self._download_webpage(url, list_id)
+
+        list_title = self._html_search_regex(r'<title>(.*?)(?:\s*-\s*BBC News)?</title>', webpage, 'list title')
+
+        pubdate = self._html_search_regex(r'"datePublished":\s*"(\d+-\d+-\d+)', webpage, 'date', default=None)
+        if pubdate:
+           pubdate = pubdate.replace('-','')
```
Comment (dstftw) [inline youtube_dl/extractor/bbcnews.py]: Use [`parse_iso8601`](https://github.com/rg3/youtube-dl/blob/master/youtube_dl/utils.py#L716-L739) to extract timestamp instead.

File: youtube_dl/extractor/bbcnews.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+        ret = []
+        # works with bbc.com/news/something-something-123456 articles
+        matches = re.findall(r"data-media-meta='({[^']+})'", webpage)
+        if not matches:
+           # stubbornly generic extractor for {json with "image":{allvideoshavethis},etc}
+           # in http://www.bbc.com/news/video_and_audio/international
+           matches = re.findall(r"({[^{}]+image\":{[^}]+}[^}]+})", webpage)
```
Comment (dstftw) [inline youtube_dl/extractor/bbcnews.py]: Use `_html_search_regex/_search_regex` with list of regex patterns instead.

File: youtube_dl/extractor/bbcnews.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        if not matches:
+           raise ExtractorError('No video found', expected=True)
+
+        for ent in matches:
+            jent = self._parse_json(ent,list_id)
+
+            programme_id = jent.get('externalId',None)
```
Comment (dstftw) [inline youtube_dl/extractor/bbcnews.py]: You don't need to specify `None` it's default.

File: youtube_dl/extractor/bbcnews.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+        ret = []
+        # works with bbc.com/news/something-something-123456 articles
+        matches = re.findall(r"data-media-meta='({[^']+})'", webpage)
+        if not matches:
+           # stubbornly generic extractor for {json with "image":{allvideoshavethis},etc}
+           # in http://www.bbc.com/news/video_and_audio/international
+           matches = re.findall(r"({[^{}]+image\":{[^}]+}[^}]+})", webpage)
+        if not matches:
+           raise ExtractorError('No video found', expected=True)
```
Comment (dstftw) [inline youtube_dl/extractor/bbcnews.py]: After you rewrite aforementioned regexes with `_search_regex` you won't need this anymore.

**Comment by dstftw** [comment]
I would also prefer placing both extractors to a single source file and finally renaming it to `bbc.py`.

**Comment by atomicdryad** [comment]
This says it has merge conflicts but it merges fine?

**Comment by dstftw** [comment]
Thanks, merged.

File: youtube_dl/extractor/bbc.py [hunk 7]
```diff
@@ -377,3 +381,177 @@ def _real_extract(self, url):
...
+                        sval['id'] = skey
+                        jsent.append(sval)
+
+        if len(jsent) == 0:
+            # stubbornly generic extractor for {json with "image":{allvideoshavethis},etc}
+            # in http://www.bbc.com/news/video_and_audio/international
+            # prone to breaking if entries have sourceFiles list
+            jsent = map(
+                lambda m: self._parse_json(m, list_id),
+                re.findall(r"({[^{}]+image\":{[^}]+}[^}]+})", webpage)
```
Comment (dstftw) [inline youtube_dl/extractor/bbc.py]: Could you provide more info on that scenario? I was unable to find any example URL that wouldn't be covered by previous patterns.

**Pull request merged**